File: Runners/JavaCodeRunner/src/main/java/csc/coderunner/JavaFileBuilder.java
package csc.coderunner;

import java.util.List;

/**
 * Generates a java file containing a single class.
 */
public class JavaFileBuilder
{
	private StringBuilder sb = new StringBuilder();
	private int indentLevel = 0;
	private int linesAdded = 0;
	
	/**
	 * Creates a builder.
	 * @param classesToImport The classes to import.
	 * @return The builder.
	 */
	public static JavaFileBuilder createBuilder(List<String> classesToImport)
	{
		JavaFileBuilder builder = new JavaFileBuilder();
		
		for (String classToImport : classesToImport)
		{
			builder.addImport(classToImport);
		}
		
		return builder;
	}
	
	/**
	 * Adds an import to the java file.
	 * @param classToImport The classes to import.
	 * @return The builder.
	 */
	private JavaFileBuilder addImport(String classToImport)
	{
		sb.append(String.format("import %s;\n", classToImport));
		linesAdded++;
		
		return this;
	}
	
	/**
	 * @return Returns the number of lines added so far.
	 */
	public int getNumLinesAdded()
	{
		return linesAdded;
	}
	
	/**
	 * Outputs a { character, and increases the indentation level.
	 * @return The builder.
	 */
	public JavaFileBuilder beginScope()
	{
		addLine("{");
		indentLevel++;
		
		return this;
	}
	
	/**
	 * Decreases the indentation level.
	 * @return The builder.
	 */
	public JavaFileBuilder endScope()
	{
		indentLevel--;
		addLine("}");
		
		return this;
	}
	
	/**
	 * Adds a line, indented by the current indentation level.
	 * @param line The line to add.
	 * @return The builder.
	 */
	public JavaFileBuilder addLine(String line, Object... args)
	{
		for (int i = 0; i < indentLevel; i++)
		{
			sb.append("\t");
		}
		
		sb.append(String.format(line, args));
		sb.append("\n");
		linesAdded++;
		
		return this;
	}
	
	/**
	 * Adds a blank line (not indented).
	 * @return The builder.
	 */
	public JavaFileBuilder addBlankLine()
	{
		sb.append("\n");
		linesAdded++;
		
		return this;
	}
	
	/**
	 * Adds each line in the given string, indented by the current 
	 * indentation level.
	 * @param lines The string containing the lines to add.
	 * @return The builder.
	 */
	public JavaFileBuilder addLines(String lines)
	{
		for (String line : lines.split("\\r?\\n"))
		{
			addLine(line);
		}
		
		return this;
	}
	
	/**
	 * @return The full contents of the file.
	 */
	public String getFileContents()
	{
		return sb.toString();
	}
}

File: Services/src/CSClassroom/CSClassroom.Service/Assignments/QuestionGeneration/JavaFileBuilder.cs
using System.Collections.Generic;
using System.Text;

namespace CSC.CSClassroom.Service.Assignments.QuestionGeneration
{
	/// <summary>
	/// Generates a java file containing a single class.
	/// </summary>
	public class JavaFileBuilder
	{
		/// <summary>
		/// The number of lines added.
		/// </summary>
		public int LinesAdded { get; private set; }

		/// <summary>
		/// The string builder that contains the contents of the file to generate.
		/// </summary>
		private readonly StringBuilder _sb = new StringBuilder();

		/// <summary>
		/// The current indent level.
		/// </summary>
		private int _indentLevel;

		/// <summary>
		/// Outputs a { character, and increases the indentation level.
		/// </summary>
		public JavaFileBuilder BeginScope(string scopeOpen)
		{
			AddLine(scopeOpen);
			_indentLevel++;

			return this;
		}

		/// <summary>
		/// Decreases the indentation level, and outputs a } character.
		/// </summary>
		public JavaFileBuilder EndScope(string scopeClose)
		{
			_indentLevel--;
			AddLine(scopeClose);

			return this;
		}

		/// <summary>
		/// Adds a line, indented by the current indentation level.
		/// </summary>
		public JavaFileBuilder AddLine(string line)
		{
			for (int i = 0; i < _indentLevel; i++)
			{
				_sb.Append("\t");
			}

			_sb.Append(line);
			_sb.Append("\n");
			LinesAdded++;

			return this;
		}

		/// <summary>
		/// Adds a blank line.
		/// </summary>
		public JavaFileBuilder AddBlankLine()
		{
			_sb.Append("\n");
			LinesAdded++;

			return this;
		}

		/// <summary>
		/// Adds each line in the given string, indented by the current 
		/// indentation level.
		/// </summary>
		public JavaFileBuilder AddLines(string lines)
		{
			return AddLines(lines.Split('\n'));
		}

		/// <summary>
		/// Adds each line in the given string, indented by the current 
		/// indentation level.
		/// </summary>
		public JavaFileBuilder AddLines(IEnumerable<string> lines)
		{
			foreach (string line in lines)
			{
				AddLine(line);
			}

			return this;
		}

		/// <summary>
		/// Returns the full contents of the file.
		/// </summary>
		public string GetFileContents()
		{
			return _sb.ToString();
		}
	}
}
